core/serialization/procedures.ts

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {IParameterModel} from '../interfaces/i_parameter_model.js';
+import {IProcedureModel} from '../interfaces/i_procedure_model.js';
+import type {ISerializer} from '../interfaces/i_serializer.js';
+import {ObservableProcedureModel} from '../procedures/observable_procedure_model.js';
+import {ObservableParameterModel} from '../procedures/observable_parameter_model.js';
+import * as priorities from './priorities.js';
+import * as serializationRegistry from './registry.js';
+import type {Workspace} from '../workspace.js';
+
+
+/**
+ * Representation of a procedure data model.
+ */
+export interface State {
+  id: string, name: string, returnTypes: string[]|null,
+      parameters?: ParameterState[],
+}
+
+/**
+ * Representation of a parameter data model.
+ */
+export interface ParameterState {
+  id: string, name: string, types?: string[],
+}
+
+/** A method that constructs an IProcedureModel. */
+type ProcedureModelFactory = (workspace: Workspace, name: string, id: string) =>
```

Comment:
We discussed offline attempting to use generics instead of the factory methods, so let me know the results of that experiment or when this is ready to re-review!